### PR TITLE
Fix scalar handling of unfold.

### PR DESCRIPTION
**Pull request by gchanan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28597 Kill _th_zero binding, just use a simple native function instead.
* #28596 Kill THTensor_(match), which isn't used.
* #28595 Write out some set_ overloads instead of relying on code binding generation.
* #28514 Kill zero_dim_tensor_only codegen, it's not used anymore.
* #28513 Remove zero_dim_tensors only from _th_masked_fill_.
* #28512 Remove zero_dim_tensors only from _th_index_fill_.
* #28511 Kill _th_fill binding, which isn't used anymore.
* #28510 unfold: turn off device_guard
* #28475 Port TH/THC implementation of unfold to ATen.
* **#28462 Fix scalar handling of unfold.**

Unfold is implemented in TH (as _th_unfold), and uses the standard scalar checks.  That means, even though torch.tensor(5).unfold(dim=0, size=1, step=1) should produce:
torch.tensor([5]), it actually produces torch.tensor(5) because the scalar_check infers it's a scalar.

We can fix this by just turning off the scalar_check.

Code change example:
Previously:
```
>>> torch.tensor(5).unfold(dimension=0, size=1, step=1)
tensor(5)
```

Now:
```
>>> torch.tensor(5).unfold(dimension=0, size=1, step=1)
tensor([5])
```

Differential Revision: [D18074671](https://our.internmc.facebook.com/intern/diff/D18074671)

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -10069,6 +10069,15 @@ def test_unfold_all_devices_and_dtypes(self, device):
                 x = torch.randint(5, (0, 1, 3, 0), dtype=dt, device=device)
                 self.assertEqual((0, 1, 1, 0, 3), x.unfold(2, 3, 2).shape)
 
+    def test_unfold_scalars(self, device):
+        x = torch.tensor(0.5, device=device)
+        # unfold on a 0-dimensional tensor should always return a 1-d dimensional
+        # tensor of shape [size] (i.e., the second parameter to unfold)
```
Comment (albanD) [inline test/test_torch.py]: Does it actually work if you do `x.unfold(0, 2, 1)` ?
The only valid values for size are 0 and 1 here right?

**Review by albanD — APPROVED** [review]
LGTM

File: test/test_torch.py (hunk 1 above)
Comment (gchanan) [reply test/test_torch.py]: Now, it will give you a error.
